ckan/lib/helpers.py

```diff
@@ -2688,3 +2688,66 @@ def sanitize_id(id_):
     ValueError.
     '''
     return str(uuid.UUID(id_))
+
+
+@core_helper
+def compare_pkg_dicts(original, new, old_activity_id):
+    '''
+    Takes two package dictionaries that represent consecutive versions of
+    the same dataset and returns a list of detailed & formatted summaries of
+    the changes between the two versions. original and new are the two package
+    dictionaries. The function assumes that both dictionaries will have
+    all of the default package dictionary keys, and also checks for additional
+    keys added by extensions and custom fields added by the user in the web
+    interface.
+
+    Returns a list of dictionaries, each of which corresponds to a change
+    to the dataset made in this revision. The dictionaries each contain a
+    string indicating the type of change made as well as other data necessary
+    to form a detailed summary of the change.
+    '''
+    from changes import _check_metadata_changes, _check_resource_changes
+    change_list = []
+
+    new_pkg = {
+        u'pkg_id': new['id'],
+        u'name': new['name'],
+        u'title': new['title']
+    }
+
+    _check_metadata_changes(change_list, original, new, new_pkg)
+
+    _check_resource_changes(change_list, original, new, new_pkg,
+                            old_activity_id)
+
+    # if the dataset was updated but none of the fields we check were changed,
+    # display a message stating that
+    if len(change_list) == 0:
+        change_list.append({u'type': 'no_change'})
+
+    return change_list
+
+
+@core_helper
+def activity_list_select(pkg_activity_list, current_activity_id):
+    '''
+    Builds an HTML formatted list of options for the select lists
+    on the "Changes" summary page.
+    '''
+    select_list = []
+    for activity in pkg_activity_list:
+        entry = render_datetime(activity['timestamp'],
+                                with_hours=True,
+                                with_seconds=True)
+        if activity['id'] == current_activity_id:
+            select_list.append(
+                "<option value=\"" + activity['id'] +
+                "\" selected>" + entry + "</option>"
```

Comment:
again, building HTML is a red flag. I've done an alternate version here: https://github.com/ckan/ckan/commit/c973f229d which you can cherry-pick